packages/zapp/console/src/components/Executions/Tables/test/NodeExecutionsTable.test.tsx

```diff
@@ -1,474 +1,198 @@
-import {
```

Comment:
Have too many comments on this one, so decided to give this file a revamp. Main concerns are: 1) avoid duplication and lines inflation by reusing input mocks and defining them at the top; 2) please avoid using standard query selectors with RTL - they have plenty built in tools that make tests more readable, and a nice cheatsheet, which I suggested you take a peek on Friday.
```
import { render, waitFor } from '@testing-library/react';
import { NodeExecutionDetailsContextProvider } from 'components/Executions/contextProvider/NodeExecutionDetails';
import {
  NodeExecutionsByIdContext,
  NodeExecutionsRequestConfigContext,
} from 'components/Executions/contexts';
import { basicPythonWorkflow } from 'mocks/data/fixtures/basicPythonWorkflow';
import { noExecutionsFoundString } from 'common/constants';
import { mockWorkflowId } from 'mocks/data/fixtures/types';
import { insertFixture } from 'mocks/data/insertFixture';
import { mockServer } from 'mocks/server';
import { RequestConfig } from 'models/AdminEntity/types';
import { NodeExecutionPhase } from 'models/Execution/enums';
import * as React from 'react';
import { dateToTimestamp } from 'common/utils';
import { QueryClient, QueryClientProvider } from 'react-query';
import { createTestQueryClient } from 'test/utils';
import { NodeExecutionsTable } from '../NodeExecutionsTable';

jest.mock('components/Workflow/workflowQueries');
const { fetchWorkflow } = require('components/Workflow/workflowQueries');

jest.mock('components/common/DetailsPanel', () => ({
  DetailsPanel: jest.fn(({ children }) => <div data-testid="details-panel">{children}</div>),
}));

jest.mock('components/Executions/Tables/NodeExecutionRow', () => ({
  NodeExecutionRow: jest.fn(({ children, execution }) => {
    return (
      <div data-testid="node-execution-row">
        <div data-testid="node-execution-col-id">{execution?.id?.nodeId}</div>
        <div data-testid="node-execution-col-phase">{execution?.closure?.phase}</div>
        {children}
      </div>
    );
  }),
}));

const selectedExecution = null;
const setSelectedExecution = jest.fn();

const node1 = {
  id: 'node1',
  scopedId: 'n1',
  type: 4,
  name: 'Node 1',
  nodes: [],
  edges: [],
};
const node2 = {
  id: 'node2',
  scopedId: 'n2',
  type: 4,
  name: 'Node 2',
  nodes: [],
  edges: [],
};
const initialNodes = [node1, node2];

const nodeExecution1 = {
  n1: {
    closure: {
      createdAt: dateToTimestamp(new Date()),
      outputUri: '',
      phase: NodeExecutionPhase.SUCCEEDED,
    },
    id: {
      executionId: { domain: 'domain', name: 'name', project: 'project' },
      nodeId: 'node1',
    },
    inputUri: '',
    scopedId: 'n1',
  },
};
const nodeExecution2 = {
  n2: {
    closure: {
      createdAt: dateToTimestamp(new Date()),
      outputUri: '',
      phase: NodeExecutionPhase.FAILING,
    },
    id: {
      executionId: { domain: 'domain', name: 'name', project: 'project' },
      nodeId: 'node2',
    },
    inputUri: '',
    scopedId: 'n2',
  },
};

describe('Executions > Tables > NodeExecutionsTable', () => {
  let queryClient: QueryClient;
  let requestConfig: RequestConfig;
  let fixture: ReturnType<typeof basicPythonWorkflow.generate>;
  let nodeExecutionsById = {};

  beforeEach(() => {
    requestConfig = {};
    queryClient = createTestQueryClient();
    fixture = basicPythonWorkflow.generate();
    insertFixture(mockServer, fixture);
    fetchWorkflow.mockImplementation(() => Promise.resolve(fixture.workflows.top));
  });

  const renderTable = (props) =>
    render(
      <QueryClientProvider client={queryClient}>
        <NodeExecutionsRequestConfigContext.Provider value={requestConfig}>
          <NodeExecutionDetailsContextProvider workflowId={mockWorkflowId}>
            <NodeExecutionsByIdContext.Provider value={nodeExecutionsById}>
              <NodeExecutionsTable {...props} />
            </NodeExecutionsByIdContext.Provider>
          </NodeExecutionDetailsContextProvider>
        </NodeExecutionsRequestConfigContext.Provider>
      </QueryClientProvider>,
    );

  it('renders empty content when there are no nodes', async () => {
    const { queryByText, queryByTestId } = renderTable({
      initialNodes: [],
      selectedExecution,
      setSelectedExecution,
    });

    await waitFor(() => queryByText(noExecutionsFoundString));
    expect(queryByText(noExecutionsFoundString)).toBeInTheDocument();
    expect(queryByTestId('node-execution-row')).not.toBeInTheDocument();
  });

  it('renders NodeExecutionRows with proper nodeExecutions', async () => {
    nodeExecutionsById = {
      ...nodeExecution1,
      ...nodeExecution2,
    };
    const { queryAllByTestId } = renderTable({
      initialNodes,
      selectedExecution,
      setSelectedExecution,
    });

    await waitFor(() => queryAllByTestId('node-execution-row'));

    expect(queryAllByTestId('node-execution-row')).toHaveLength(initialNodes.length);
    const ids = queryAllByTestId('node-execution-col-id');
    expect(ids).toHaveLength(initialNodes.length);
    expect(ids[0]).toHaveTextContent(initialNodes[0].id);
    expect(ids[1]).toHaveTextContent(initialNodes[1].id);
    const phases = queryAllByTestId('node-execution-col-phase');
    expect(phases).toHaveLength(initialNodes.length);
    expect(phases[0]).toHaveTextContent(NodeExecutionPhase.SUCCEEDED.toString());
    expect(phases[1]).toHaveTextContent(NodeExecutionPhase.FAILING.toString());
  });

  it('renders future nodes with UNDEFINED phase', async () => {
    nodeExecutionsById = { ...nodeExecution1 };
    const { queryAllByTestId } = renderTable({
      initialNodes,
      selectedExecution,
      setSelectedExecution,
    });

    await waitFor(() => queryAllByTestId('node-execution-row'));

    expect(queryAllByTestId('node-execution-row')).toHaveLength(initialNodes.length);
    const ids = queryAllByTestId('node-execution-col-id');
    expect(ids).toHaveLength(initialNodes.length);
    expect(ids[0]).toHaveTextContent(initialNodes[0].id);
    expect(ids[1]).toHaveTextContent(initialNodes[1].id);
    const phases = queryAllByTestId('node-execution-col-phase');
    expect(phases).toHaveLength(initialNodes.length);
    expect(phases[0]).toHaveTextContent(NodeExecutionPhase.SUCCEEDED.toString());
    expect(phases[1]).toHaveTextContent(NodeExecutionPhase.UNDEFINED.toString());
  });
});
```

Reply:
@james-union have you read though this file carefully?
1. We don't need to waitFor `container` if expect other elements that we gonna test
2. You still declare `selectedExecution` and `setSelectedExecution` in every test, while all of them are the same
3. We can have multiple expectations out of the test, ie both length and text content in my examples. Why? When tests break, it will be easier and faster to fix them, knowing what specific area of the test doesn't match existing expectations